src/backend/common/epoch.cpp

```diff
@@ -0,0 +1,123 @@
+    // order of fetching smallest_epoch, largest_epoch and max_cid is VERY
+    // IMPORTANT
+    auto smallest_epoch = txn_manager.GetSmallestEpochCleanedCid();
+    // get the largest epoch that can be cleaned
+    auto largest_epoch = txn_manager.GetCurrentEpochId();
+    auto max_cid = txn_manager.GetMaxCommittedCid();
+    if (max_cid == MAX_CID) {
+      // no transaction is running, i.e. all epochs until largest epoch can be
+      // cleaned
+      // update the largest epoch according to the max_epochs_per_thread
+      assert(largest_epoch >= smallest_epoch);
+      if (largest_epoch - smallest_epoch > max_epochs_per_thread) {
+        largest_epoch = smallest_epoch + max_epochs_per_thread;
+      }
+      // Perform CAS on the smallest epoch. This ensures that only one
+      // thread (the one that wins the CAS) cleans one epoch
+      if (txn_manager.PerformEpochCAS(smallest_epoch, largest_epoch)) {
+        // iterate through all the epochs from smallest to largest and clean
+        // them
+        for (; smallest_epoch < largest_epoch; smallest_epoch++) {
+          Epoch *e = txn_manager.GetEpoch(smallest_epoch);
+          if (e == nullptr) {
+            continue;
+          }
+          gc_manager.PerformGC(e);
+          // if cleaned some epoch other than self, clean it
+          // if cleaned self, signal the calling function (EndTransaction) using
+          // retval
+          // the calling function will delete the epoch in that case
+          txn_manager.EraseEpoch(smallest_epoch);
+          if (e != this) {
+            delete e;
+          } else {
+            retval = true;
+          }
+        }
+        return retval;
+      } else {
+        // lost the CAS, simply return
+        return retval;
+      }
+    } else if (smallest_epoch != INVALID_CID) {
+      // There are some running threads, we can clean tuples upto max_cid
+      max_cid++;
+      // adjust max_cid according to max_epochs_per_thread
+      if (max_cid - smallest_epoch > max_epochs_per_thread) {
+        max_cid = smallest_epoch + max_epochs_per_thread;
+      }
+      // perform CAS so that only one thread does this
+      if (txn_manager.PerformEpochCAS(smallest_epoch, max_cid)) {
+        assert(max_cid >= smallest_epoch);
+        // clean the epochs from [smallest_epoch, max_cid), because we
+        // incremented max_cid above
+        for (; smallest_epoch < max_cid; smallest_epoch++) {
+          Epoch *e = txn_manager.GetEpoch(smallest_epoch);
+          if (e == nullptr) {
+            continue;
+          }
+          // perform GC of this epoch's possibly free list
+          gc_manager.PerformGC(e);
+          txn_manager.EraseEpoch(smallest_epoch);
+          if (e != this) {
+            // delete self reference afterwards, handled by setting the retval
+            delete e;
+          } else {
+            retval = true;
+          }
+        }
+        return retval;
+      } else {
+        return retval;
+      }
```

Comment:
Line 91-117 is essentially the same as line 57-86, except for that you use max_cid rather than largest_epoch here. So it would be better if you refract them and create a function, say PerformGC(....) to remove duplicated code.